### ISSUE TITLE
No longer functioning on Disney+

### ISSUE THREAD

**Issue by PoobaSama** [dvlden/uw-pub#15]
**Describe the bug**
Used to work fine on disney plus on chrome but is no longer working. Tried on firefox as well and it's the same issue.

**Screenshots/Videos**

![image](https://github.com/dvlden/uw-pub/assets/71660512/0a0f9baf-f005-44ed-937d-f9e0b7a6ddb5)

**Desktop (please complete the following information):**
 - OS: [e.g. Windows 10]
 - Browser [e.g. Chrome]
 - Version [e.g. Version	10.0.19045 Build 19045]

**Comment by nnx0r** [comment]
Anything at all? 
Does it show banner inside of the video, to show which is the currently active mode?

It works for me on macOS, but not for this specific show. It does not upscale/stretch all the way on my screen, just partially. Not sure if it's that aspect ratio limitation or Disney+ limitation. 

However, I tried using custom scaling X/Y with normal mode and that works just fine. 
Please let me know, about those tiny details.

EDIT:
Also, please try some other movies/series and let me know. Cause this feedback is very important to me!

**Comment by PoobaSama** [comment]
Tried a couple of shows and it doesn't seem to work either. Works fine on other sites. 

No banner inside of video. Also tried the scaling with normal but it didn't seem to change anything.

**Comment by nnx0r** [comment]
No banner, can only mean that something is broken. Either with the platform or something is conflicting with the extension itself.

I'll try tonight on VM using Windows. Keep you updated!

**Comment by Puppies-On-Acid** [comment]
Any progress with this. Still not working with Disney+

**Comment by Somenamethatworks** [comment]
Running Firefox on Windows 10.
This extension was working fine for me with Disney+ up until last night when I cleared cookies, cache. and did a Norton file cleanup. 
Now the extension will not work with Disney+.    :(

**Comment by LeonNamowitz** [comment]
Same thing here Win10 and Chrome Version 122.0.6261.112

**Comment by leonavis** [comment]
Same on Arch Linux and FF 123.0.1

**Comment by petran79** [comment]
Same on Windows 10 and Firefox. Used to work previously.

**Comment by nnx0r** [comment]
Sorry for keeping you out in the dark. I'm on vacation until monday, and I did not have a chance to test it yet! 

Will get back to all of you on Monday, with the news. Hoping it's an easy fix, otherwise, I plan to change the implementation, but that may take some time.

Thank you all for reporting, I really appreciate every single one of you.

**Comment by leonavis** [comment]
Just checked and it seems to work with Family Guy but not with "The Bad Batch". Strange...

**Comment by nnx0r** [comment]
Hello everybody! 
I have been able to test this today on VM and also on a physical device. Apparently in both cases it no longer works for me too. It only seems to be working on macOS for some reason, but I will assume they enabled some kind of A/B testing and enabled this new feature only for some devices or accounts.

After my investigation, I have came to conclusion that Disney+ has built now a custom player using [Web Components](https://developer.mozilla.org/en-US/docs/Web/API/Web_components). For those that do not know what that means, it basically means that they made a native `<video />` element, wrapped within their custom element for whatever reasons they had. One of which could be - prevent the extensions to do crazy things with video player.

UltraWideo does not currently support custom Web Components, as it doesn't know how to find them. They can be named whatever, for example `<asdf />` and it can be a video player inside. If video streaming platforms had some kind of consistent naming, for example `<disney-player />` and `<netflix-player />` _(Netflix doesn't use custom component, it's just example)_, then it would be easy to know it's custom video element, by suffix `player`. But of course, they do not comply with it.

Luckily, Disney's new Web Component is named `<disney-web-player />`, so my idea here is to extend the utility that finds the video element and to allow it to also try to find custom Web Components that ends with `-player`, cause I do not want to hardcode full element name for various reasons:

- They could rename it at some point
- It would be able to find only Disney's custom player

I will keep you guys updated on the progress. As soon as I figure out how to approach it the best.

TLDR: It's fixable and fix is coming!

---

### Update 2
I came up with a plan to traverse through all elements on the page and find the Custom Web Component that potentially contains `video` element inside of it. 

However, Disney Plus came up with one more overhead. Every mouse movement, they keep replacing all classes with the ones that their player is officially using. 

This means, that they strip UW classes that are required for the extension to fully function.
I will try two things:

1. Prevent mouse-move event from bubbling to their page (effectively preventing them to do what they're doing)
2. Change internal logic of extension to not rely on classes at all

I'll see what works and what is simpler to achieve for quick fix, to make my users happy. 
Afterwards, I'll see what is more performant for another update.

--- 

### Update 3
Apparently it's not just that. Injected styles are completely ignored by Web Components, because they are shadowed. Unfortunately, that means that even with the change where I dropped the use of classes, it still does nothing to the video player itself.

I have to figure out the good way to inject styles into the shadow-root as well. It's a lot more work than I anticipated.

**Comment by dvlden** [comment]
I came up with a plan to traverse through all elements on the page and find the Custom Web Component that potentially contains `video` element inside of it. 

However, Disney Plus came up with one more overhead. Every mouse movement, they keep replacing all classes with the ones that their player is officially using. 

This means, that they strip UW classes that are required for the extension to fully function.
I will try two things:

1. Prevent mouse-move event from bubbling to their page (effectively preventing them to do what they're doing)
2. Change internal logic of extension to not rely on classes at all

I'll see what works and what is simpler to achieve for quick fix, to make my users happy. 
Afterwards, I'll see what is more performant for another update.

**Comment by nnx0r** [comment]
Finally some news from my end!
I was able to make it work for both Chromium and Firefox. Here's one quick preview from Chrome on Windows.

https://github.com/dvlden/uw-pub/assets/3460368/af5796c7-9aba-48a9-aef8-a9fd4e6acc57

I'm still not 100% sure if it's good enough and does it break anything else. So I will not release the update tonight. Will take some time for further manual testing.

Stay tuned!

**Comment by nnx0r** [comment]
Just a quick update. I have tested it thoroughly and I'm happy with it. 
Polishing some release notes and changelogs before release.

I can't guarantee, but I think I will release this tomorrow.

**Comment by nnx0r** [comment]
Just pushed the v3.1.0 to Chrome Webstore and Firefox Add-ons. Awaiting review.
Reviews usually take sometime, especially on Firefox store, because of the long manual review.

You can read release notes here, but you'll be greeted with this page automatically, once extension is approved and your browser updates it.

[v3.1.0 Changelog](https://uw.wtf/changelog?version=3.1.0)

Let me know if issue has been resolved, when you get the update. Thanks!

**Comment by leonavis** [comment]
Works for me. Great work, thank you. :)

**Comment by kyounger** [comment]
Is there a way to download the xpi ahead of that publishing by Firefox?

**Comment by nnx0r** [comment]
@kyounger I am unfamiliar with how would that work. If you are familiar or can send me an guide that I can follow, I'll be happy to provide it for you.

**Comment by kyounger** [comment]
Ahhh, I don't think I realized until now that this extension isn't open source. That would explain why the builds and releases aren't available here on GH. I'll just wait until the update goes through in the FF extension repo.

**Comment by nnx0r** [comment]
Yeah, closed source since v3.0.0 - cause of planned pro features. 
I found a Firefox guide... Signing the extension right now and as soon as it's done, I'll upload it here.

Keep in mind tho, that once you install it, you need to watch out for the release, cause I don't think it will auto-update (maybe it will). But in case it won't, you'd have to delete this beta and install latest.

Stay tuned, not sure how long signing process is going to take. I have pretty fast machine, but it's been over 2 minutes already. 

EDIT:
Nevermind, it still requires review from their team, even if it's going to be self-hosted. I do not get it...

**Comment by kyounger** [comment]
FF Extension still look to be in approval, but Chrome extension in Brave works again for Disney+ now.

**Comment by nnx0r** [comment]
@kyounger Yup. Firefox is so slow, it's sad. Not sure what is happening with them. Instead of leading the industry, they are at least 7 years behind Chromium on technology improvements.

**Comment by nnx0r** [comment]
I wonder how Apple will act on this behalf when I release UltraWideo for Safari. Hope they are fast like Chrome Web Store.

**Comment by nnx0r** [comment]
![image](https://github.com/dvlden/uw-pub/assets/3460368/abd0d817-d48d-4d39-b4fa-63fa844c00d3)

**Comment by petran79** [comment]
Something tells me that they do this in order to hide their awful video streaming quality. With monitor resolution set at 3440x1440 and 100 Mbps connection, it should be able to approach ultrawide 4K video content resolution (3840x1608).
Using this addon and stretching at ultrawide, quality is awful. Unless player refuses to stream 4K resolution version because of monitor resolution
https://i.ibb.co/Q8zj3dM/20240315-213026.jpg

This picture is from the 4K BD version rip.
https://i.ibb.co/WKdFkML/20240315-213046.jpg

**Comment by GwenBD94** [comment]
@dvlden 
Windows 11 Pro 23H2 build 22631.3291
Firefox 123.0.1 64 bit
UltraWideo 3.1.0

Still doesn't work on Disney+. Opened the new The Marvels movie in Disney+, opened the addon menu by clicking it's icon, switched back and forth between stretch normal and upscale, nothing had any affect on the video. Tried again on princess diaries 2, no dice.

**Comment by petran79** [comment]
> @dvlden Windows 11 Pro 23H2 build 22631.3291 Firefox 123.0.1 64 bit UltraWideo 3.1.0
> 
> Still doesn't work on Disney+. Opened the new The Marvels movie in Disney+, opened the addon menu by clicking it's icon, switched back and forth between stretch normal and upscale, nothing had any affect on the video. Tried again on princess diaries 2, no dice.

in my case it would only stretch X. Used upscale instead and played with the values and picture was zoomed. Then used upscale 1x1 again and ratio appeared with correct proportions

**Comment by nnx0r** [comment]
@GwenBD94 Do you see the banner inside of the video, appearing when you are cycling through the modes or upscaling/downscaling?

@petran79 yeah I noticed this on Disney too. Sometimes, depending on the video's AR and screen size, their video doesn't want to go all the way upscaled and requires the use of Scaling option in combination.

Scale **X: 1** & **Y: 1** should be defaults by the way. In case you change them while watching something, you should always revert it back to 1/1 after you're done, cause some other video aspect ratio or platform may work properly without additional upscaling/downscaling.

I am on 14" laptops (Macbook and Razer) and I tried multiple shows, including the ones that @GwenBD94 has mentioned. It works great for me, with additional setting of scale Y that required a bump to `1.1` in combination with `Upscale` mode.

> I'd appreciate if you guys could record a short video, showing me what are the settings, what is the show, is it specific to an show, what is your screen size and resolution. Much appreciated!

**Comment by GwenBD94** [comment]
@dvlden I don't see any type of banner in the video no. I would love to record you a short video, or even take screenshots, if not for DRM causing the entire screen to just be a black blob every time i open any recording or screenshot tools. 
![image](https://github.com/dvlden/uw-pub/assets/99558676/d309fbfc-2ce9-4b7d-a195-fbbc246ab359)


Screen is a curved (1000r iirc?) 34" monoprice dark matter 3440X1440 21:9 144hz

EDIT: decided to test it on another site and it looks like your plugin just doesn't work for me at all. rip. Hulu and Nebula also had no effect whatsoever

**Comment by petran79** [comment]
you can use your smartphone too

**Comment by nnx0r** [comment]
@GwenBD94 What @petran79 said. Use smartphone to record your screen.

I'd love to resolve this issue for you. If it doesn't work on Hulu either (not sure what is Nebula, will check) then it's definitely a problem.

Can you try using it in private mode (incognito) where you will only allow UW and not any other extensions, so we can exclude the option that it conflicts with something else.

Would you be up for a TeamViewer session or software alike, where I could access your computer and test it out?

Best regards!

**Comment by GwenBD94** [comment]
https://photos.google.com/photo/AF1QipNH2snB_rqkdikS-5yQE88i0WHWJJXjntfUj5iS

@dvlden did private mode with nothing but your extension and 1password, still nothing.

**Comment by nnx0r** [comment]
Can't see your pic. That should be fine, I use 1Pass too.

How about that TeamViewer session?

![image](https://github.com/dvlden/uw-pub/assets/3460368/da219e1b-5908-4cfe-914f-26caf6021821)

**Comment by GwenBD94** [comment]
try https://photos.app.goo.gl/wfNZ5Dd4R4eaATWf6

**Comment by nnx0r** [comment]
@GwenBD94 Oh I see. I may not have been clear with the instructions on the onboarding page, perhaps. You have to put the video in the fullscreen so that the extension knows what it should interact with. 

Either fast double-click anywhere on the video or click that square icon on the bottom right.

Let me know...

EDIT:
Extension have multiple shortcuts, so you don't have to open up popup at all. Google: Extension shortcuts configuration in Chrome/Firefox for more details.

**Comment by GwenBD94** [comment]
Ah then it's not quite what I was looking for for my usecase then. Thanks anyways

**Comment by nnx0r** [comment]
No worries @GwenBD94 !
Would love to know what's your use-case?

**Comment by nnx0r** [comment]
@GwenBD94 If you don't mind, open a Discussion regarding your use-case. I'd really appreciate your input.

I'd like to close this issue as the issue has been resolved for Disney+ in 3.1.0, among with some other platforms that use custom player component.

**Comment by mariusgherasie** [comment]
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
Adespinoza, @LindsayRichman, @lfarlima, @rocco, @HorIv001, @lluisagusti, @faint32, @shihanjun19, @chrishna1, @tkersey, @code2cash, @YiHerngOng, @zhousongsong, @Abnerfang, @manguro-code, @wandun, @superern, @johneris, @lizhifei2010, @EribertoLopez, @softfn, @Don2025, @sinwoobang, @AndrewKevin, @ayonliu, @FlepNL, @aemoe, @crmitchelmore, @johnnymn, @Potat-Cat, @allenlin90, @erikiado, @guitorioadar, @zengyang0428, @epogrebnyak, @0xdevalias, @whidy, @Hishmad, @vinhphuctadang, @xinming365, @ihoro, @chenAdmin, @bychannel, @AmirDiafi, @c0r1y-6a0, @SouthForm, @xavieresp1, @TeslowMusic, @thomas-gale, @Parables
